blogs/2020/02/14/optimizing-ci.md

```diff
@@ -0,0 +1,117 @@
+
+```yml
+- task: 1ESLighthouseEng.PipelineArtifactCaching.RestoreCacheV1.RestoreCache@1
+  inputs:
+    keyfile: keyfile
+    targetfolder: target
+    vstsFeed: $(ArtifactFeed)
+    platformIndependent: true
+```
+
+When caching node modules, it’s logical to use `package-lock.json` files as a cache key: when this file changes, we must invalidate the cache. When caching compilation outputs, the entire source code must act as a cache key. To simplify things, we decided to use the HEAD commit as the cache key—a new commit will inevitably create a new cache entry. This works fine for our purposes since a single build, despite running across build agents, always runs over a single commit.
+
+Another missing feature was the ability to create multiple caches per build job. We now found ourselves juggling two caches (node modules, compilation) with no way to address each cache individually. The caching tasks output an environment variable called `CacheRestored` that can be used to optimistically skip build tasks. This environment variable works great in builds that interact with a single cache, but not so great with multiple caches—leaving us to wonder which cache `CacheRestored` is referring to. Once again, we worked on a way to add an optional `alias` parameter to the task. For example:
+
+```yml
+- task: 1ESLighthouseEng.PipelineArtifactCaching.RestoreCacheV1.RestoreCache@1
+  inputs:
+    keyfile: "yarn.lock"
+    targetfolder: "node_modules"
+    vstsFeed: "$(ArtifactFeed)"
+    alias: "Packages"
+
+- script: |
+    yarn install
+  displayName: Install Dependencies
+  condition: ne(variables['CacheRestored-Packages'], 'true')
+```
+
+Here, an alias of `Packages` is appended to the environment variable output, allowing us to cache NPM packages and compilation output in a single build job. We were finally deduplicating a lot of the CI work that can now be executed only once and shared across platform-specific agents.
+
+There was still room for one final optimization, given a specific use case: build resubmissions. We sometimes must retrigger builds on previously built commits since tests might be flaky or some agents might randomly fail. Ideally, the shared agent would not restore or recompile the common code but simply defer to the platform-dependent agents to perform their work. The issue we noticed was that the compilation cache packages were massive and restoring them would take around 8 minutes—all for nothing since the shared agent would simply yield control if that cache existed. We addressed this issue by implementing a `dryRun` feature to allow us to check for a cache package’s existence without restoring it—effectively cutting 8 minutes off our build resubmissions. Using the `dryRun` parameter looks like this:
```

Comment:
Make 'we' more specific

>We sometimes must retrigger builds on previously built commits since tests might be flaky or some agents might randomly fail.

In VS Code we...